x/autopilot/keeper/ibc.go

```diff
@@ -0,0 +1,162 @@
+package keeper
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	transfertypes "github.com/cosmos/ibc-go/v7/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+)
+
+// Deserializes the acknowledgement and returns a bool indicating whether it was successful or was an ack error
+func (k Keeper) CheckAcknowledgementStatus(ctx sdk.Context, acknowledgementBz []byte) (success bool, err error) {
```

Comment:
Why do we need so much new custom logic in `ibc.go` to handle the fallback address? Would have expected a lot of this to come out of the box (e.g. checking the status of an ack)

OnAck and OnTimeout seem like a lot of code too. Can/should we make those more modular to re-use tried-and-tested helpers?

Possible we need all this, but if possible we'd minimize the new LOC in ibc middleware, as it's particularly sensitive.

Reply:
I think this function here is the only non-business logic'y code in this file (i.e. the only code that would have overlap with something else). We do have a function in icacallbacks that does something similar and was the blueprint for this, but it had some additional functionality that wasn't needed, so I thought a thinned down version would make more sense

Reply:
Is there anything specifically in OnAck or OnTimeout that you think should be modularized? Imo, each line of each function is either calling a keeper function, calling a helper function, or doing something specific to the fallback case (e.g. bank send), so I don't really see how it could be modularized more, but open to ideas if you have some!

Reply:
can we implement this as a callback? 

it seems like most of the repeated logic is because this approach doesn't use `icacallbacks` (which should really be called `ibccallbacks`)

Reply:
ah yeah I guess we probably could. I was thinking it would need to be part of the middleware stack, but I think we changed that recently so where it's no longer in the stack

It's not obvious to me that it would be much cleaner, cause the callback module itself comes with it's own handful of boilerplate that you need to add to the module, plus the datastructure wrapper is a bit bloated/messy for something as simple as storing down a single address. But I can look into it more tomorrow

Reply:
good point, we'd be swapping out some boilerplate code for new boilerplate code. my thinking was that the boilerplate we currently have is more sensitive than the `icacallbacks` boilerplate, but there might be other costs associated with using `icacallbacks` here

Reply:
Yeah it's a good point. Fwiw, I think the keeper boilerplate code is probably the most bullet proof code in cosmos - it's replicated across basically every data structure on every chain and easily verifiable via unit tests.

So it's a question of adding net new, but simple code (this PR), versus wiring up a more complex solution that doesn't require net new code (ibcallbacks). 

It's not obvious to me which makes more sense but I'll go through the exercise tomorrow of seeing what callbacks would look like and I'm sure the better solution will show though!

Reply:
@sampocs Are we taking the callback approach? About to review but will pause if we're going to rewrite as a callback

Reply:
Sorry haven't had a chance to look into it yet. I think you can still review it in the meantime and just focus on the business logic instead of the keepers/CheckAckStatus function. The business logic should be identical with both, it's just a difference between using keeps or wiring up callbacks

Reply:
chatted offline, tabling the callback approach for now given the complexity involved. We can re-evaluate the best approach for our middleware stack in the next few months

marking this resolved!